pkg/scheduler/server/internal/conn.go

```diff
@@ -15,57 +15,136 @@ package internal
 
 import (
 	"context"
+	"errors"
+	"io"
 	"sync"
 
 	schedulerv1pb "github.com/dapr/dapr/pkg/proto/scheduler/v1"
 )
 
+// conn represents a single connection bidirectional stream between the
+// scheduler and the client. conn manages sending triggered jobs to the client,
+// and receiving job process results from the client. Jobs are sent serially
+// via a channel as gRPC does not support concurrent sends.
+// conn tracks the inflight jobs and acks them when the client sends back the
+// result, releasing the job triggering.
 type conn struct {
-	wg       sync.WaitGroup
-	closeCh  chan struct{}
-	streamer *streamer
-	ch       chan *schedulerv1pb.WatchJobsResponse
-}
+	pool  *Pool
+	jobCh chan *schedulerv1pb.WatchJobsResponse
 
-type streamer struct {
-	lock sync.RWMutex
-	subs map[uint32]chan struct{}
-}
+	// idx is the uuid of a triggered job. We can use a simple counter as there
+	// are no privacy/time attack concerns as this counter is scoped to a single
+	// client.
+	idx uint64
+
+	// inflight tracks the jobs that have been sent to the client but have not
+	// been acked yet.
+	inflight map[uint64]chan struct{}
 
-func (s *streamer) add(id uint32) chan struct{} {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-	ch := make(chan struct{})
-	s.subs[id] = ch
-	return ch
+	lock    sync.RWMutex
+	closeCh chan struct{}
 }
 
-func (s *streamer) handleResponse(id uint32) {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-	if ch, ok := s.subs[id]; ok {
-		ch <- struct{}{}
+// newConn creates a new connection and starts the goroutines to handle sending
+// jobs to the client and receiving job process results from the client.
+func (p *Pool) newConn(req *schedulerv1pb.WatchJobsRequestInitial, stream schedulerv1pb.Scheduler_WatchJobsServer, uuid uint64) *conn {
+	conn := &conn{
+		pool:     p,
+		closeCh:  make(chan struct{}),
+		inflight: make(map[uint64]chan struct{}),
+		jobCh:    make(chan *schedulerv1pb.WatchJobsResponse, 10),
 	}
-	delete(s.subs, id)
+
+	p.wg.Add(2)
+
+	go func() {
+		defer p.wg.Done()
+		defer p.remove(req, uuid)
+		defer close(conn.closeCh)
+		for {
+			select {
+			case job := <-conn.jobCh:
+				if err := stream.Send(job); err != nil {
+					log.Warnf("Error sending job to connection: %v", err)
```

Comment:
Should we log the jobName or any conn details here that may be helpful for debugging?

Reply:
Can you add back in the comment about adding to an orphaned go routine if the send fails, so we don't forget later on?

Reply:
Which comment are you referring to?

Reply:
[I think during some refactoring the comment got removed](https://github.com/cicoyle/dapr/blob/c492f19f0a3b6b625b626dfcd52767adca2ac0b0/pkg/scheduler/server/server.go#L354-L357)